### ISSUE TITLE
Mutual TLS

### ISSUE THREAD

**Issue by pellepelster** [hivemq/hivemq-mqtt-client#324]
Hi,

I am currently struggling with setting up a mutual TLS connection to a MQTT server. Coming from paho-mqtt where I was able to inject a PEMSocketFactoryFactory into the mqtt client lib, I am wondering how this would be achievable with hivemq-mqtt-client. I found ways to inject a KeyManagerFactory and/or TrustManager but it is not clear to me how to implement mutual TLS.

Any hints would be highly appreciated.

**Comment by SgtSilvio** [comment]
Hi @pellepelster,

**What are Trust- and KeyManagerFactory:**
The TrustManagerFactory is used for authentication of the remote side -> the server.
The KeyManagerFactory is used for authentication of the local side -> the client.

**What is the content:**
If you are using self-signed server certificates you have to add the CA certificates to the TrustManagerFactory.
For using client authentication you have to store the private key and the client certificate in the KeyManagerFactory.

**How can I build the Trust- and KeyManagerFactory:**
Keys and certificates are stored in `KeyStore`s.
You use KeyStores to initialize Trust- and KeyManagerFactories:
```java
KeyStore keyStore = ...
kmf = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm());
kmf.init(keyStore, privateKeyPassword);

KeyStore trustStore = ...
tmf = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
tmf.init(trustStore);
```
I guess you have a pem file for the client key and certificate.
Then you need to load the private key and certificat and add them to a KeyStore:
```java
KeyStore keyStore = KeyStore.getInstance(KeyStore.getDefaultType());
keyStore.load(null, null);
keyStore.setKeyEntry("key", privateKey, privateKeyPassword, certificateChain);
```

I hope this information helps.
It would be good, if you could provide more information about how your key material is stored and if you are using self-signed certificates.

**Comment by clive-jevons** [comment]
Hi @pellepelster 

Here's an integration of the HiveMQ Client in another framework I'm working on:

https://github.com/bmwcarit/joynr/blob/1.8.3/java/messaging/mqtt/hivemq-mqtt-client/src/main/java/io/joynr/messaging/mqtt/hivemq/client/HivemqMqttClientFactory.java

Towards the end of the file you can see how I went about configuring the SSL key and trust stores. As it's a framework integration it's probably a bit overkill for your purposes, but hopefully you'll be able to glean some useful insights from it :)

Cheers,
C

**Comment by waikontse** [comment]
@pellepelster

Hi, I've had the same question as you, with the mTLS configuration and all and with the same setup.
So the certificate and private key is a pem file. I've attached a working example of how I load the files into a keystore and a truststore at the gist https://gist.github.com/waikontse/b6cb58277170ab0f9fac865744787984.

= Note
Important stuff to be aware of, the private key is in pem format and pkcs8. I used bouncycastle PEMParser to do some plumbing for me. If you don't want to pull in bouncycastle dependency, you would have to remove the "-----BEGIN PRIVATE KEY-----" and "-----END PRIVATE KEY-----" from your pkcs8 pem encoded private key file and use Base64.Decoder to get the expected raw bytes array for your private key.
